developer/src/kmc-package/src/compiler/package-version-validation.ts

```diff
@@ -0,0 +1,139 @@
+import { KmpJsonFile, CompilerCallbacks, KpsFile, KmxFileReader, KMX } from '@keymanapp/common-types';
+import { CompilerMessages } from './messages.js';
+
+export class PackageVersionValidation {
+
+  constructor(private callbacks: CompilerCallbacks) {}
+
+  /**
+   * Verifies version information in corresponding keyboards and updates kmpJson
+   * metadata as the version information can be out of sync in the .kps file
+   * after update a contained keyboard.
+   * @param kpsFilename
+   * @param kps
+   * @param kmp
+   * @returns
+   */
+  public validateAndUpdateVersions(kpsFilename: string, kps: KpsFile.KpsFile, kmp: KmpJsonFile.KmpJsonFile) {
+    const followKeyboardVersion = kps.options?.followKeyboardVersion !== undefined;
+
+    if(followKeyboardVersion) {
+      if(!this.checkFollowKeyboardVersion(kps, kmp)) {
+        return false;
+      }
+    }
+
+    let result = true;
+
+    if(!kmp.keyboards) {
+      // Lexical model packages don't have version metadata
+      return true;
+    }
+
+    // We now know we have at least one keyboard in the package
+
+    for(let keyboard of kmp.keyboards) {
+      result = this.updateKeyboardVersionFromKmx(kpsFilename, kmp, keyboard) && result;
```

Comment:
Question: should we just fail fast here and simplify the logic (eliminate `result` var)? However, this may mean multiple recompiles for users as subsequent errors with the keyboards would not be shown in the compiler messages until the first error is fixed. Of course, that may be of minimal advantage here as any failure in validation will still cancel the build anyway.

Reply:
I'm personally a fan of stockpiling all appropriate error messages and reporting them together, rather than requiring multiple rebuilds just to get one error message at a time.

For people just starting out with keyboard development, they may otherwise feel like "oh, I only did one thing wrong" just to get hammered repeatedly with one "another" thing they did wrong, with no end in sight.  That can be pretty demotivating to a newbie.  Having a list of things may be daunting at first, but you can visualize your progress better as the number of errors goes down, tackling "easier" things first - which may not be the first reported error.

Reply:
That _is_ true.  I have... "fond" memories of getting 100+ compile errors when first learning C++.  Because of a single missing semi-colon, of course.